rustls/src/tls13/mod.rs

```diff
@@ -199,6 +209,76 @@ impl MessageDecrypter for Tls13MessageDecrypter {
     }
 }
 
+struct RingAeadFactory(&'static ring::aead::Algorithm);
+
+impl RingAeadFactory {
+    fn encrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageEncrypter> {
+        Box::new(Tls13MessageEncrypter {
+            enc_key: aead::LessSafeKey::new(aead::UnboundKey::new(self.0, key.as_ref()).unwrap()),
+            iv,
+        })
+    }
+
+    fn decrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageDecrypter> {
+        Box::new(Tls13MessageDecrypter {
+            dec_key: aead::LessSafeKey::new(aead::UnboundKey::new(self.0, key.as_ref()).unwrap()),
+            iv,
+        })
+    }
+
+    fn key_len(&self) -> usize {
+        self.0.key_len()
+    }
+}
+
+struct AeadChacha20Poly1305(RingAeadFactory);
```

Comment:
Would suggest making `Aead` a suffix rather than a prefix for these type names.

This abstraction layering feels a bit odd. I think `RingAeadFactory` actually wants to be a trait with one method (the ring `Algorithm`) and three provided methods?

Reply:
i think this is in a much better place now: have deleted the three `AeadChacha20Poly1305` types which weren't adding much

Reply:
oh, actually they were providing separate impls of `Tls13AeadAlgorithm` which differed for `extract_keys`. hmm